### ISSUE TITLE
Cannot reproduce stage3 training results.

### ISSUE THREAD

**Issue by haomengz** [allenai/WildDet3D#23]
Hi,

I have tried to train the model for stage3 using the following command, starting from stage-2 checkpoint:
```
vis4d fit --config configs/training/stage3_mix_box_point_text_ft.py \
--gpus 4 \
--ckpt ckpt/wilddet3d_stage2_alldata_12e_v1.0.pt
```

I notice that `importance_sampling` is no longer an argument of `affine_invariant_local_loss()` in the current MoGe [repo](https://github.com/microsoft/MoGe/blob/07444410f1e33f402353b99d6ccd26bd31e469e8/moge/train/losses.py#L112), so I simply remove this argument in `affine_invariant_local_loss()`:
https://github.com/allenai/WildDet3D/blob/1b8aa52b6ff3f00d0ebfa07175efc0c0c440964a/wilddet3d/depth/lingbot_backend.py#L754
https://github.com/allenai/WildDet3D/blob/1b8aa52b6ff3f00d0ebfa07175efc0c0c440964a/wilddet3d/depth/lingbot_backend.py#L764

Everything else is the same. After 3 epochs finetuning, I got AP_3D of 19.82 for the `last.ckpt` when evaluating with text prompt and without depth.

As mentioned in #20, it seems the provided stage2 checkpoint is already well-trained, and the performance is close to the stage3 checkpoint. 3 epochs fine-tuning is hurting the performance. Could you please recheck the uploaded stage2 checkpoint? Please let me know if there is any problem of my reproducing procedures. Thanks！

**Comment by weikaih04** [comment]
yes I will take a look recently

**Comment by weikaih04** [comment]
Hi @haomengz, thanks for the detailed report — and sorry for the slow reply.

A few things to clarify:

**1. Effective batch / lr mismatch is the dominant issue**

Our paper Stage 3 was trained on **4 nodes × 8 GPU = 32 GPUs**, giving an effective batch of 128 (4 samples/GPU × 32 GPUs × 1 accumulate). The `lr=1e-4` in the config is calibrated for that effective batch.

Your run used `--gpus 4`, so the effective batch is 16 — **8× smaller** — which makes `lr=1e-4` effectively 8× too high and destabilises training on top of the well-trained Stage 2 checkpoint.

The cleanest fix on 4 GPUs is to recover effective batch 128 with gradient accumulation:

```bash
vis4d fit --config configs/training/stage3_mix_box_point_text_ft.py \
    --gpus 4 \
    --ckpt ckpt/wilddet3d_stage2_alldata_12e_v1.0.pt \
    --config.params.accumulate_grad_batches=8
```

(4 GPUs × 4 samples × accumulate 8 = effective batch 128.)

**2. `importance_sampling=False` removal is a secondary concern**

Upstream MoGe (`microsoft/MoGe`) removed the `importance_sampling` kwarg and now always uses importance sampling internally. Our paper run was with `importance_sampling=False` (uniform sampling). The gradient signal does differ, but the lr/batch mismatch above is the dominant effect — removing the kwarg as you did is the right call given the current MoGe API.

**3. About what Stage 3 is supposed to do**

Stage 3 is **not primarily a metric-boost stage**. Its main purpose is to adapt the Stage 2 model to all three prompt formats — text, box geometry, and point geometry — via the mixed 5-mode collator. The text-prompt AP after Stage 3 is expected to be roughly on par with Stage 2 on Omni3D; the real value of Stage 3 shows up in box / point prompt evaluation and in better generalisation on WildDet3D-Bench. If your Stage 2 already evaluates close to the paper number under text prompts, you can stop there for that specific metric.

**4. If you only need the paper weights**

The HF release `wilddet3d_alldata_all_prompt_v1.0.pt` is exactly the Stage 3 checkpoint from the paper (I verified state_dict equality vs my local training output). For exact paper numbers, just download and evaluate that file rather than re-running Stage 3.

Let me know if anything is still off after using the accumulation fix.

**Comment by haomengz** [comment]
Thanks for the clarification. I have verified with a subset of data and the accumulation fix seems to be effective.
